### PR TITLE
Derive isResolved dynamically from subscribers resolved breach data classes (MNTOR-2871)

### PR DESCRIPTION
**Pull request by flozia**
# References: 
Jira: [MNTOR-2871](https://mozilla-hub.atlassian.net/browse/MNTOR-2871)

# Description

Instead of storing `isResolved` as part of the breach resolution in the `SubscriberRow` we should rather dynamically derive if a breach is resolved. We can do that by checking the breaches the user is affected by and the resolved data classes we store along with the subscriber. This allows for the option to conditionally mark breaches as resolved — namely relevant for this PR: Making SSN breaches part of a breach resolution depending on location.

# How to test

Prerequisites and the steps to follow are described in [MNTOR-2871](https://mozilla-hub.atlassian.net/browse/MNTOR-2871).

[MNTOR-2871]: https://mozilla-hub.atlassian.net/browse/MNTOR-2871?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/utils/subscriberBreaches.test.ts [hunk 1]
```diff
@@ -223,7 +272,51 @@ describe("getSubBreaches", () => {
       >
     ).mockResolvedValueOnce(breachesWithOneResolved);
 
-    const subBreaches = await getSubBreaches(subscriber, []);
+    const subBreaches = await getSubBreaches(subscriber, [], "nl");
+    expect(subBreaches.length).toEqual(1);
+    expect(subBreaches[0].isResolved).toBe(true);
+  });
+
+  it("returns that a breach containing a SSN is resolved for US users", async () => {
```
Comment (Vinnl) [inline src/utils/subscriberBreaches.test.ts:280]: Shouldn't we want to verify that a breach with an SSN is marked as _un_resolved for US users, but resolved for non-US users?

**Review by Vinnl — COMMENTED** [review]
It looks good to me and I'm very happy to see this change. But I'd like Joe or Rob to approve as well, since we'll stop recording `isResolved` in the database from this point on, so in that sense this takes effect right away (even if it doesn't involve a slow migration).

File: src/db/tables/subscribers.js [hunk 2]
```diff
@@ -277,35 +277,10 @@ async function setAllEmailsToPrimary (subscriber, allEmailsToPrimary) {
 /* c8 ignore stop */
 
 /**
- * OBSOLETE, preserved for backwards compatibility
```
Comment (mansaj) [inline src/db/tables/subscribers.js:280]: Thanks! I'll have to make a ticket to start cleaning up some of these obsolete functions as well as columns in our `subscriber` row. 

Some columns that are no longer being used include:
```
fx_newsletter
breaches_last_shown
breaches_resolved
waitlists_joined
breach_stats
...
```

and all the corresponding util funcs

Comment (mansaj) [reply src/db/tables/subscribers.js:280]: Added: https://mozilla-hub.atlassian.net/browse/MNTOR-2972 

FYI

File: src/app/functions/universal/breach.ts [hunk 3]
```diff
@@ -49,3 +52,17 @@ export const SecurityRecommendationDataTypes = {
...
+  dataClassesAffected: DataClassEffected[],
+  resolvedDataClasses: Array<HibpBreachDataTypes[keyof HibpBreachDataTypes]>,
+) {
+  return (
+    dataClassesAffected.every((dataClassAffected) => {
+      const dataClassAffectedKey = Object.keys(
+        dataClassAffected,
+      )[0] as (typeof BreachDataTypes)[keyof typeof BreachDataTypes];
+      return resolvedDataClasses.includes(dataClassAffectedKey);
+    }) || false
```
[the remark this answers is not in the document]
Comment (flozia) [reply src/app/functions/universal/breach.ts]: Not needed and has been a fallback for backward compatibility with V1 — removed in [b130578](https://github.com/mozilla/blurts-server/pull/4212/commits/b130578597fb532d57abedf23a2106d81eec726c).

File: src/utils/subscriberBreaches.test.ts (hunk 1 above)
Comment (flozia) [reply src/utils/subscriberBreaches.test.ts:280]: Added a test case that also verify that breaches that include SSN are unresolved for US users in [b130578](https://github.com/mozilla/blurts-server/pull/4212/commits/b130578597fb532d57abedf23a2106d81eec726c).

File: src/apiMocks/mockData.ts [hunk 4]
```diff
@@ -112,11 +112,14 @@ export function createRandomBreach(
...
   const isResolved = options.isResolved ?? faker.datatype.boolean();
+  const dataClassesEffected = options.dataClassesEffected ?? [];
+  const resolvedDataClasses = isResolved ? dataClasses : [];
+
   return {
     addedDate: options.addedDate ?? faker.date.recent(),
     breachDate: faker.date.recent(),
     dataClasses: dataClasses,
-    resolvedDataClasses: isResolved ? dataClasses : [],
+    resolvedDataClasses: resolvedDataClasses,
```
Comment (mansaj) [inline src/apiMocks/mockData.ts:122]:
```suggestion
    resolvedDataClasses,
```

File: src/app/(proper_react)/(redesign)/(authenticated)/user/(dashboard)/dashboard/fix/leaked-passwords/[type]/page.tsx [hunk 5]
```diff
@@ -51,15 +55,12 @@ export default async function LeakedPasswords({
       subscriberEmails={subscriberEmails}
       type={type}
       data={{
-        countryCode: getCountryCode(headers()),
+        countryCode: countryCode,
```
Comment (mansaj) [inline src/app/(proper_react)/(redesign)/(authenticated)/user/(dashboard)/dashboard/fix/leaked-passwords/[type]/page.tsx:58]:
```suggestion
        countryCode,
```

File: src/app/functions/server/getUserBreaches.ts [hunk 6]
```diff
@@ -95,10 +95,16 @@ export async function getUserBreaches({
  * NOTE: new function to replace getUserBreaches
  *
  * @param user
+ * @param user.user
+ * @param user.countryCode
  */
-export async function getSubscriberBreaches(
-  user: Session["user"],
-): Promise<SubscriberBreach[]> {
+export async function getSubscriberBreaches({
```
Comment (mansaj) [inline src/app/functions/server/getUserBreaches.ts:101]: is there an advantage of passing in an object of `{user, countryCode}` vs. two params `(user, countryCode)`?

Comment (flozia) [reply src/app/functions/server/getUserBreaches.ts:101]: TS makes it easier to know what arguments to pass to a function. Generally, this is a good way to know which one a function takes. We do it in a couple of places, but I don’t think we follow a consistent pattern.

File: src/apiMocks/mockData.ts (hunk 4 above)
Comment (flozia) [reply src/apiMocks/mockData.ts]: Removed redundant key/value pairs in [dd3200d](https://github.com/mozilla/blurts-server/pull/4212/commits/dd3200d38555c0993d42adb94d364cff3670b585).

File: src/app/(proper_react)/(redesign)/(authenticated)/user/(dashboard)/dashboard/fix/leaked-passwords/[type]/page.tsx (hunk 5 above)
Comment (flozia) [reply src/app/(proper_react)/(redesign)/(authenticated)/user/(dashboard)/dashboard/fix/leaked-passwords/[type]/page.tsx]: Removed redundant key/value pairs in [dd3200d](https://github.com/mozilla/blurts-server/pull/4212/commits/dd3200d38555c0993d42adb94d364cff3670b585).

**Review by mansaj — APPROVED** [review]

**Comment by yawayl** [comment]
Some one must be the muskateer steal real in my thoughts. Thank you/

**Pull request merged**